File: backend/src/Operum.API/Controllers/ViewsController.cs
using Microsoft.AspNetCore.Mvc;
using Operum.API.Controllers.Base;
using Operum.Model.DTOs.Views.Requests;
using Operum.Service.Interfaces;

namespace Operum.API.Controllers
{
    [ApiController]
    [Route("api/trackers/{trackerId}/[controller]")]
    public class ViewsController(IViewsService viewsService) : BaseController
    {
        [HttpPost]
        public async Task<IActionResult> CreateView([FromBody] CreateViewDto view, [FromRoute] string trackerId)
        {
            return GetApiResponse(await viewsService.CreateView(trackerId, view));
        }

        [HttpGet("{viewId}")]
        public async Task<IActionResult> GetView([FromRoute] string trackerId, [FromRoute] string viewId)
        {
            return GetApiResponse(await viewsService.GetView(trackerId, viewId));
        }

        [HttpGet]
        public async Task<IActionResult> GetViewList([FromRoute] string trackerId)
        {
            return GetApiResponse(await viewsService.GetViewList(trackerId));
        }

        [HttpPut("{viewId}")]
        public async Task<IActionResult> UpdateView([FromRoute] string trackerId, [FromRoute] string viewId, [FromBody] UpdateViewDto view)
        {
            return GetApiResponse(await viewsService.UpdateView(trackerId, viewId, view));
        }

        [HttpDelete("{viewId}")]
        public async Task<IActionResult> DeleteView([FromRoute] string trackerId, [FromRoute] string viewId)
        {
            return GetApiResponse(await viewsService.DeleteView(trackerId, viewId));
        }

        [HttpPut("reorder")]
        public async Task<IActionResult> ReorderViews([FromRoute] string trackerId, [FromBody] ReorderViewsDto reorderViews)
        {
            return GetApiResponse(await viewsService.ReorderViews(trackerId, reorderViews));
        }
    }
}

File: frontend/src/features/views/api/viewsController.ts
import api from "../../../shared/api/api";
import { ApiResponse } from "../../../shared/types/ApiResponse";
import { CreateViewDto } from "../types/requests/CreateViewDto";
import { UpdateViewDto } from "../types/requests/UpdateViewDto";
import { ViewDto } from "../types/ViewDto";

export const viewsController = {
    getView: async (
        trackerId: string,
        viewId: string
    ): Promise<ApiResponse<ViewDto>> => {
        return await api.get(`trackers/${trackerId}/views/${viewId}`);
    },
    getViewList: async (trackerId: string): Promise<ApiResponse<ViewDto[]>> => {
        return await api.get(`trackers/${trackerId}/views`);
    },
    setDefaultView: async (
        trackerId: string,
        viewId?: string
    ): Promise<ApiResponse> => {
        return await api.put(`/trackers/${trackerId}/default-view`, null, {
            params: { defaultViewId: viewId },
        });
    },
    createView: async (
        trackerId: string,
        values: CreateViewDto
    ): Promise<ApiResponse> => {
        return await api.post(`/trackers/${trackerId}/views`, values);
    },
    updateView: async (
        trackerId: string,
        viewId: string,
        values: UpdateViewDto
    ): Promise<ApiResponse> => {
        return await api.put(`/trackers/${trackerId}/views/${viewId}`, values);
    },
    deleteView: async (
        trackerId: string,
        viewId: string
    ): Promise<ApiResponse> => {
        return api.delete(`trackers/${trackerId}/views/${viewId}`);
    },
    updateViewOrder: async (
        trackerId: string,
        viewIds: string[]
    ): Promise<ApiResponse> => {
        return await api.put(`/trackers/${trackerId}/views/reorder`, { viewIds });
    },
};
